emote/models/model_env.py

```diff
@@ -0,0 +1,197 @@
+            initial_obs_batch (torch.Tensor): a batch of initial observations.
+            len_rollout (int): the max length of the model rollout
+        """
+        self._timestep = 0
+        self._len_rollout = len_rollout
+
+        assert len(initial_obs_batch.shape) == 2  # batch, obs_dim
+        self._current_obs = to_tensor(initial_obs_batch).to(self.dynamic_model.device)
+        self._init_obs = torch.clone(self._current_obs)
+
+    def step(
+        self,
+        actions: np.ndarray,
+    ) -> tuple[Tensor, Tensor | None, Tensor, dict[str, Tensor]]:
+        """Steps the model environment with the given batch of actions.
+
+        Arguments:
+            actions (np.ndarray): the actions for each "episode" to rollout.
+                Shape must be batch_size x dim_actions. If a np.ndarray is given, it's
+                converted to a torch.Tensor and sent to the model device.
+
+        Returns:
+            (Union[tuple, dict]): contains the predicted next observation, reward, done flag.
+            The done flag and rewards are computed using the termination_fn and
+            reward_fn passed in the constructor. The rewards can also be predicted
+            by the model.
+        """
+        assert len(actions.shape) == 2  # batch, action_dim
+        with torch.no_grad():
+            actions = torch.from_numpy(actions).to(self.device)
+            (next_observs, pred_rewards,) = self.dynamic_model.sample(
+                action=actions,
+                observation=self._current_obs,
+                rng=self.rng,
+            )
+            rewards = (
+                pred_rewards
+                if self.reward_fn is None
+                else self.reward_fn(actions, next_observs)
+            )
+            dones = self.termination_fn(next_observs)
+
+            info = {"terminated": torch.zeros(dones.shape)}
+            self._timestep += 1
+            if self._timestep >= self._len_rollout:
+                info["terminated"] += 1.0
+            self._current_obs = torch.clone(next_observs)
+            return next_observs, rewards, dones, info
+
+    def dict_step(
+        self,
+        actions: dict[AgentId, DictResponse],
+    ) -> tuple[dict[AgentId, DictObservation], dict[str, float]]:
+        """The function to step the Gym-like model with dict_action.
+
+        Arguments:
+            actions (dict[AgentId, DictResponse]): the dict actions.
+
+        Returns:
+            (tuple[dict[AgentId, DictObservation], dict[str, float]]): the predicted next dict observation,
+            reward, and done flag.
+        """
+        batched_actions = np.stack(
+            [actions[agent].list_data["actions"] for agent in self._agent_ids]
+        )
+        next_obs, rewards, dones, info = self.step(batched_actions)
+        new_agents = []
+        results = {}
+        terminated = info["terminated"]
+
+        for env_id, (done, term) in enumerate(zip(dones, terminated)):
+            if done or term:
```

Comment:
Nit: I'd not use terminated/term because below it reads like a bug that term isn't linked with terminal. Human pattern-seeking and all that. I'd at least add a comment that `terminated` means the time ran out, not that it hit a terminal state.

Reply:
Fair point :)
I renamed it to reached_max_len